editoast/src/views/timetable/track_occupancy_utils.rs

```diff
@@ -0,0 +1,269 @@
+use chrono::DateTime;
+use chrono::Duration;
+use chrono::Utc;
+use core_client::simulation::ReportTrain;
+use schemas::infra::TrackOffset;
+use schemas::primitives::PositiveDuration;
+use schemas::train_schedule::OperationalPointIdentifier;
+use schemas::train_schedule::PathItemLocation;
+
+use crate::views::path::path_item_cache::PathItemCache;
+use crate::views::path::pathfinding::PathfindingResult;
+use crate::views::path::projection::PathProjection;
+use crate::views::path::projection::TrackLocationFromPath;
+use crate::views::projection::find_index_upper;
+use crate::views::projection::linear_interpolate;
+use crate::views::timetable::simulation;
+use crate::views::timetable::simulation::SimulationResponseSuccess;
+
+#[derive(Debug, Clone)]
+pub struct TrackOccupancyResult {
+    pub track_section: String,
+    pub time_begin: DateTime<Utc>,
+    pub duration: PositiveDuration,
+}
+
+/// Common function to compute track occupancy on a specific path item
+pub fn compute_track_occupancy_on_path_item(
+    train_schedule: &schemas::TrainSchedule,
+    path_item_id: &str,
+    path_item_positions: &[u64],
+    path_projection: &PathProjection,
+    operational_point_track_offsets: &[TrackOffset],
+    report_train: &ReportTrain,
+) -> Option<TrackOccupancyResult> {
+    let schedule_item = train_schedule
+        .schedule
+        .iter()
+        .find(|schedule| schedule.at.0 == path_item_id);
+
+    let stop_duration = schedule_item
+        .and_then(|item| item.stop_for.clone())
+        .unwrap_or_default();
+
+    let index = train_schedule
+        .path
+        .iter()
+        .position(|path| path.id.0 == path_item_id)?;
+
+    let position = path_item_positions[index];
+    let time = report_train.path_item_times[index];
+
+    let track_section = match path_projection.get_location(position) {
+        TrackLocationFromPath::One(track_offset) => operational_point_track_offsets
+            .iter()
+            .find(|to| *to == &track_offset)?
+            .track
+            .to_string(),
+        TrackLocationFromPath::Two(track_offset, track_offset1) => operational_point_track_offsets
+            .iter()
+            .find(|to| *to == &track_offset)
+            .or(operational_point_track_offsets
+                .iter()
+                .find(|to| *to == &track_offset1))?
+            .track
+            .to_string(),
+    };
+
+    let time_begin = train_schedule.start_time + Duration::milliseconds(time as i64);
+
+    Some(TrackOccupancyResult {
+        track_section,
+        time_begin,
+        duration: stop_duration,
+    })
+}
+
+/// Common function to interpolate track occupancy
+pub fn interpolate_track_occupancy(
```

Comment:
This doesn't need to be public (but you might need to migrate some tests from `train_schedule.rs`).

```suggestion
fn interpolate_track_occupancy(
```

Reply:
Done.